File: AL-PA05/pa05_fish.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

vector<int> flip(const vector<int>& arr, int i, int j) {
    vector<int> tmp = arr;
    reverse(tmp.begin() + i, tmp.begin() + j + 1);
    for (int k = i; k <= j; ++k) {
        tmp[k] = -tmp[k];
    }
    return tmp;
}

bool check_one(const vector<int>& arr) {
    vector<int> target(arr.size());
    for (int i = 0; i < arr.size(); ++i) {
        target[i] = i + 1;
    }
    int start = min_element(arr.begin(), arr.end()) - arr.begin();
    int end = 0;
    for (int i = 1; i <= arr.size(); ++i) {
        auto it = find(arr.begin(), arr.end(), -i);
        if (it != arr.end()) {
            end = it - arr.begin();
            break;
        }
    }
    vector<int> res = flip(arr, start, end);
    return res == target;
}

bool check_continuous_subarrays(const vector<int>& arr) {
    int n = arr.size();
    int i = 0;
    vector<pair<int, int>> sub_arrays;
    while (i < n) {
        int j = i + 1;
        while (j < n && arr[j] == arr[j - 1] + 1) {
            ++j;
        }
        if (j > i) {
            sub_arrays.emplace_back(i, j);
            if (check_one(flip(arr, i, j - 1))) {
                return true;
            }
        }
        i = j;
    }
    if (check_one(flip(arr, sub_arrays[0].first, sub_arrays[1].second - 1))) {
        return true;
    }
    if (check_one(flip(arr, sub_arrays[0].first, sub_arrays[2].second - 1))) {
        return true;
    }
    if (check_one(flip(arr, sub_arrays[1].first, sub_arrays[2].second - 1))) {
        return true;
    }
    if (check_one(flip(arr, sub_arrays[1].first, sub_arrays[3].second - 1))) {
        return true;
    }
    if (check_one(flip(arr, sub_arrays[2].first, sub_arrays[3].second - 1))) {
        return true;
    }
    return false;
}

bool check_two(const vector<int>& arr) {
    return check_continuous_subarrays(arr);
}

void check(const vector<int>& arr) {
    if (check_one(arr)) {
        cout << "one\n";
    } else if (check_two(arr)) {
        cout << "two\n";
    } else {
        cout << "over\n";
    }
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);

    int n;
    cin >> n;
    vector<vector<int>> bread_tray(5, vector<int>(n));
    for (int i = 0; i < 5; ++i) {
        for (int j = 0; j < n; ++j) {
            cin >> bread_tray[i][j];
        }
    }
    for (const auto& arr : bread_tray) {
        check(arr);
    }
    return 0;
}

File: AL-PA05/pa05_fish.py
def flip(arr, i, j):
    tmp = arr[:]
    tmp[i:j + 1] = [-x for x in reversed(tmp[i:j + 1])]
    return tmp


def check_one(arr):
    target = [x + 1 for x in range(len(arr))]
    start = arr.index(min(arr))
    end = 0
    for i in range(1, len(arr) + 1):
        if -i in arr:
            end = arr.index(-i)
            break
    res = flip(arr, start, end)
    if res == target:
        return True
    return False


def check_continuous_subarrays(arr):
    n = len(arr)
    i = 0
    sub_arrays = []
    while i < n:
        j = i + 1
        while j < n and arr[j] == arr[j - 1] + 1:
            j += 1
        if j > i:
            sub_arrays.append((i, j))
            if check_one(flip(arr, i, j - 1)):
                return True
        i = j
    if check_one(flip(arr, sub_arrays[0][0], sub_arrays[1][1] - 1)):
        return True
    if check_one(flip(arr, sub_arrays[0][0], sub_arrays[2][1] - 1)):
        return True
    if check_one(flip(arr, sub_arrays[1][0], sub_arrays[2][1] - 1)):
        return True
    if check_one(flip(arr, sub_arrays[1][0], sub_arrays[3][1] - 1)):
        return True
    if check_one(flip(arr, sub_arrays[2][0], sub_arrays[3][1] - 1)):
        return True
    return False


def check_two(arr):
    if check_continuous_subarrays(arr):
        return True
    return False


def check(arr):
    if check_one(arr):
        print('one')
    elif check_two(arr):
        print('two')
    else:
        print('over')


n = int(input())
target = [x + 1 for x in range(n)]
bread_tray = []

for i in range(5):
    tmp_arr = list(map(int, input().split()))
    bread_tray.append(tmp_arr)

for arr in bread_tray:
    check(arr)
